<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>libSBOL: $title</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">libSBOL
   &#160;<span id="projectnumber">2.0.1</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Introduction</span></a></li>
      <li><a href="getting_started.html"><span>Getting&#160;Started</span></a></li>
      <li><a href="sequences.html"><span>Sequence&#160;Assembly</span></a></li>
      <li><a href="modular_design.html"><span>Biosystem&#160;Design</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="contents">
<div class="textblock"><h1><a class="anchor" id="modular_design"></a>
Biosystem Design</h1>
<p>LibSBOL can help synthetic biologists design modular systems of biochemical interactions. Following is an example of such a circuit adapted from <a href="http://www.nature.com/ncomms/2014/141106/ncomms6393/abs/ncomms6393.html">Synthesizing AND gate genetic circuits based on CRISPR-Cas9 for identification of bladder cancer cells</a>. This circuit consists of a NAND gate coupled to a NOT gate. The NAND gate turns off in the presence of specific tumor markers and tissue markers, thus its behavior is targeted to specific tissue and tumor sites in the human body. When the NAND gate is shut off, the downstream NOT gate is turned on, thus the coupled effect of the two modules will activate the apoptotic (self-destruct) protein hBAX. Thus, the entire system behaves as a targeted tumor-killer. </p><div class="image">
<img src="CRISPRTumorKiller.png" alt="CRISPRTumorKiller.png"/>
</div>
<p> A representation of this system in SBOL can be programmatically constructed. Starting at the highest level of abstraction, a tumor-killer circuit will be constructed from the NAND and NOT gates. At this stage the modules are black boxes without any internal specifications. </p><div class="fragment"><div class="line">ModuleDefinition&amp; NandGate = *<span class="keyword">new</span> ModuleDefinition(<span class="stringliteral">&quot;NandGate&quot;</span>);</div><div class="line">ModuleDefinition&amp; NotGate = *<span class="keyword">new</span> ModuleDefinition(<span class="stringliteral">&quot;NotGate&quot;</span>);</div><div class="line">ModuleDefinition&amp; TumorKiller = *<span class="keyword">new</span> ModuleDefinition(<span class="stringliteral">&quot;TumorKiller&quot;</span>);</div><div class="line">doc.add &lt; ModuleDefinition &gt; (NandGate);</div><div class="line">doc.add &lt; ModuleDefinition &gt; (NotGate);</div><div class="line">doc.add &lt; ModuleDefinition &gt; (TumorKiller);</div><div class="line">TumorKiller.assemble({ &amp;NotGate, &amp;NandGate });</div></div><!-- fragment --><p> In order to connect these black boxes, their inputs and outputs must be defined. The following process defines an interface for the modules, while their internal workings still remain abstract. First, configure the NAND gate: </p><div class="fragment"><div class="line">ComponentDefinition&amp; TissueMarker = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;TissueMarker&quot;</span>, <a class="code" href="constants_8h.html#a5eeedc63b5c9afebba2dd9e197a1851b">BIOPAX_PROTEIN</a>);</div><div class="line">ComponentDefinition&amp; TumorMarker = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;TumorMarker&quot;</span>, <a class="code" href="constants_8h.html#ad5d83dee393dbbc05f76e5cc057a0d21">BIOPAX_RNA</a>);</div><div class="line">ComponentDefinition&amp; LacI = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;LacI&quot;</span>, <a class="code" href="constants_8h.html#a5eeedc63b5c9afebba2dd9e197a1851b">BIOPAX_PROTEIN</a>);</div><div class="line">doc.add &lt; ComponentDefinition &gt; ({ &amp;TissueMarker, &amp;TumorMarker, &amp;LacI });</div><div class="line">FunctionalComponent&amp; nand_input1 = NandGate.setInput(TissueMarker);</div><div class="line">FunctionalComponent&amp; nand_input2 = NandGate.setInput(TumorMarker);</div><div class="line">FunctionalComponent&amp; nand_output = NandGate.setOutput(LacI);</div></div><!-- fragment --><p> Next configure the NOT gate. Note that LacI's ComponentDefinition was already defined above. </p><div class="fragment"><div class="line">ComponentDefinition&amp; HBax = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;HBax&quot;</span>, <a class="code" href="constants_8h.html#a5eeedc63b5c9afebba2dd9e197a1851b">BIOPAX_PROTEIN</a>);</div><div class="line">doc.add &lt; ComponentDefinition &gt; (HBax);</div><div class="line">FunctionalComponent&amp; not_input = NotGate.setInput(LacI);</div><div class="line">FunctionalComponent&amp; not_output = NandGate.setOutput(HBax);</div></div><!-- fragment --><p> Finally, these two modules can be connected with a metaphorical, yet very satisfying, "click"." </p><div class="fragment"><div class="line">nand_output.connect(not_input);</div></div><!-- fragment --> <h1><a class="anchor" id="pathway_modeling"></a>
Pathway Modeling</h1>
<p>The preceding example demonstrated how black-box modules may be connected in terms of their inputs and outputs. In order to enable computer-aided engineering (model-based design and simulation of biochemical systems), libSBOL also supports detailed description of biochemical interactions and pathways. The SBOL standard was designed to be closely interoperable with SBML (Systems Biology Markup Language) so that information about structural components can be associated with mathematical descriptions of their behavior. Let's now look at defining modules at a deeper level of scrutiny. </p><div class="image">
<img src="CRISPRTemplate.png" alt="CRISPRTemplate.png"/>
</div>
<p> Above is a representation of a template module for CRISPR gene-editing that consists of three biochemical interactions. As described previously (see <a class="el" href="sequences.html#template_design">Partial versus Template Designs</a>), template designs may be re-used in many different contexts. For example, a synthetic biologist may want to substitute many different sequences for a guideRNA depending on the actual downstream target. In this example, each component has a specifed functional role while detailed information about its sequence is to be filled in at a later time. Starting at a high level we define the template module and its three member interactions using Systems Biology Ontology (SBO) terms. In the first interaction, Cas9 protein and a guideRNA form a complex. This activated complex targets the promoter cognate to the guideRNA and disables gene expression. The third interaction is expression of the target gene which is normally on by default, but switched off in the presence of activated complex. </p><div class="fragment"><div class="line"><span class="comment">// Initialize the template CRISPR module</span></div><div class="line">ModuleDefinition&amp; CRISPRTemplate = *<span class="keyword">new</span> ModuleDefintion(<span class="stringliteral">&quot;CRISPRTemplate&quot;</span>);</div><div class="line">ComponentDefinition&amp; Cas9 = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;Cas9&quot;</span>, <a class="code" href="constants_8h.html#a5eeedc63b5c9afebba2dd9e197a1851b">BIOPAX_PROTEIN</a>);</div><div class="line">ComponentDefinition&amp; GuideRNA = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;GuideRNA&quot;</span>, <a class="code" href="constants_8h.html#ad5d83dee393dbbc05f76e5cc057a0d21">BIOPAX_RNA</a>);</div><div class="line">ComponentDefinition&amp; Cas9GuideRNAComplex = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;Cas9GuideRNAComplex&quot;</span>, <a class="code" href="constants_8h.html#abcdfbca32134be773472d9af0f825e44">BIOPAX_COMPLEX</a>);</div><div class="line">ComponentDefinition&amp; TargetPromoter = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;TargetPromoter&quot;</span>, <a class="code" href="constants_8h.html#aa08e807e9ad29c17c56e3b17f088f6cc">BIOPAX_DNA</a>);</div><div class="line">ComponentDefinition&amp; TargetGene = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;TargetGene&quot;</span>, <a class="code" href="constants_8h.html#aa08e807e9ad29c17c56e3b17f088f6cc">BIOPAX_DNA</a>);</div><div class="line">ComponentDefinition&amp; TargetProtein = *<span class="keyword">new</span> ComponentDefinition(<span class="stringliteral">&quot;TargetProtein&quot;</span>, <a class="code" href="constants_8h.html#a5eeedc63b5c9afebba2dd9e197a1851b">BIOPAX_PROTEIN</a>);</div><div class="line">doc.add &lt; ModuleDefinition &gt; (CRISPRTemplate);</div><div class="line">doc.add &lt; ComponentDefinition &gt; ({ &amp;Cas9, &amp;GuideRNA, &amp;Cas9GuideRNAComplex, &amp;TargetPromoter, &amp;TargetGene, &amp;TargetProtein });</div><div class="line"></div><div class="line"><span class="comment">// Define the internal interactions in the module</span></div><div class="line">Interaction&amp; complex_formation = CRISPRTemplate.interactions.create(<span class="stringliteral">&quot;complex_formation&quot;</span>);</div><div class="line">Interaction&amp; gene_repression = CRISPRTemplate.interactions.create(<span class="stringliteral">&quot;gene_repression&quot;</span>);</div><div class="line">Interaction&amp; gene_production = CRISPRTemplate.interactions.create(<span class="stringliteral">&quot;gene_production&quot;</span>);</div><div class="line">complex_formation.types.add(SBO_NONCOVALENT_BINDING);</div><div class="line">gene_repression.types.add(SBO_INHIBITION);</div><div class="line">gene_production.types.add(SBO_GENETIC_PRODUCTION);</div></div><!-- fragment --><p> Now let's break the Interactions down into their member species, starting with the reaction for complex formation. Note the use of SBO terms to define the role of each species in the reaction. </p><div class="fragment"><div class="line"><span class="comment">// Complex formation A + B -&gt; AB</span></div><div class="line">Participation&amp; A = complex_formation.participations.create(<span class="stringliteral">&quot;A&quot;</span>);</div><div class="line">Participation&amp; B = complex_formation.participations.create(<span class="stringliteral">&quot;B&quot;</span>);</div><div class="line">Participation&amp; AB = complex_formation.participations.create(<span class="stringliteral">&quot;AB&quot;</span>);</div><div class="line">A.roles.add(SBO_REACTANT);</div><div class="line">B.roles.add(SBO_REACTANT);</div><div class="line">AB.roles.add(SBO_PRODUCT);</div></div><!-- fragment --><p> The components which participate in an interaction must be assigned: </p><div class="fragment"><div class="line"><span class="comment">// Assign a ComponentDefinition to the reactants and products</span></div><div class="line">Cas9.participate(A);</div><div class="line">GuideRNA.participate(B);</div><div class="line">Cas9GuideRNAComplex.participate(AB);</div></div><!-- fragment --><p> Similarly, the repression and production interactions are defined: </p><div class="fragment"><div class="line"><span class="comment">// Gene repression</span></div><div class="line">Participation&amp; repressor = complex_formation.participations.create(<span class="stringliteral">&quot;repressor&quot;</span>);</div><div class="line">Participation&amp; repression_target = complex_formation.participations.create(<span class="stringliteral">&quot;repression_target&quot;</span>);</div><div class="line">repressor.roles.add(SBO_INHIBITOR);</div><div class="line">repression_target.roles.add(SBO_BINDING_SITE);</div><div class="line"></div><div class="line"><span class="comment">// Gene production</span></div><div class="line">Participation&amp; promoter = gene_production.participations.create(<span class="stringliteral">&quot;promoter&quot;</span>);</div><div class="line">Participation&amp; gene = gene_production.participations.create(<span class="stringliteral">&quot;gene&quot;</span>);</div><div class="line">Participation&amp; gene_product = gene_production.participations.create(<span class="stringliteral">&quot;gene_product&quot;</span>);</div><div class="line">promoter.roles.add(SBO_PROMOTER);</div><div class="line">gene.roles.add(SBO_GENE);</div><div class="line">gene_product.roles.add(SBO_PRODUCT);</div><div class="line"></div><div class="line">Cas9GuideRNAComplex.participate(repressor);</div><div class="line">TargetPromoter.participate(repression_target);</div><div class="line">TargetPromoter.participate(promoter);</div><div class="line">TargetGene.participate(gene);</div><div class="line">TargetProtein.participate(gene_product);</div></div><!-- fragment --> <h1><a class="anchor" id="template_override"></a>
Completing a Template Design</h1>
<p>A complete design can be realized from a template design by overriding components. In the CRISPR template above, the lacI-gRNA and LacI components override the generic guideRNA and gene product. The dashed lines here have a different meaning than they did in the tumor-killer example. There the dashed lines indicate that a module's output component is identical to another modules' input component. Here the dashed lines indicate that the template component is being substituted with the LacI protein. Thus it is possible to make many possible variant designs from a template design. To create a variant from a template design, nest the template module inside the variant module. Then mask the inner module's component with the outer module's component. </p><div class="fragment"><div class="line">ModuleDefinition&amp; CRISPRTemplate = *<span class="keyword">new</span> ModuleDefinition(<span class="stringliteral">&quot;CRISPRTemplate&quot;</span>);</div><div class="line">ModuleDefinition&amp; CRISPRVariant = *<span class="keyword">new</span> ModuleDefinition(<span class="stringliteral">&quot;CRISPRVariant&quot;</span>);</div><div class="line">doc.add&lt;ModuleDefinition&gt;({ &amp;CRISPRTemplate, &amp;CRISPRVariant });</div><div class="line">CRISPRVariant.assemble({ &amp;CRISPRTemplate });</div><div class="line">FunctionalComponent&amp; template_output = CRISPRTemplate.setOutput(TargetProtein);</div><div class="line">FunctionalComponent&amp; variant_output = CRISPRTemplate.setOutput(LacI);</div><div class="line">variant_output.mask(template_output);</div></div><!-- fragment --> </div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Fri Aug 12 2016 16:57:39 for libSBOL by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.11
</small></address>
</body>
</html>
